src/event-processor/event-processor.service.ts

```diff
@@ -0,0 +1,223 @@
+          durable: true,
+        });
+      });
+
+      this.logger.log('Connected to RabbitMQ.');
+      return true;
+    } catch (error) {
+      this.logger.error('Failed to connect to RabbitMQ', error);
+      return false;
+    }
+  }
+
+  /**
+   * Initializes the event consuming for a queue
+   * @param queue the queue identifier
+   */
+  private initializeConsumer(queue: string) {
+    this.channel.consume(
+      queue,
+      (msg) => {
+        if (msg && this.processingAllowed[queue]) {
+          const message = msg.content.toString();
+          this.logger.debug(
+            `[${queue}] Processing message: ${message} [${this.messageCounts[queue] + 1}/${this.maxMessages}]`,
+          );
+          this.channel.ack(msg);
+          this.messageCounts[queue]++;
+          // randomise timeout to simulate processing
+          const delay = Math.floor(Math.random() * this.processingTime);
+          this.logger.log(`[${queue}] Sending Event Update after [${delay}]s`);
+          // Wait for the set amount of time
+          setTimeout(() => {
+            this.buildEventUpdate(queue, message);
+          }, delay);
+
+          if (this.messageCounts[queue] >= this.maxMessages) {
+            this.logger.debug(
+              `[${queue}] Maximum message count reached. Pausing processing until reset.`,
+            );
+            this.processingAllowed[queue] = false;
+          }
+        }
+      },
+      { noAck: false },
+    );
+  }
+
+  /**
+   * Builds the DTOs for the updated event for each queue
+   * @param queueName The string identifier of the queue
+   * @param msg The received message
+   */
+  private buildEventUpdate(queueName: string, msg: any) {
+    // flip successfull event with given sucess rate
+    const successfull: boolean = Math.random() < this.successRate;
+    switch (queueName) {
+      case 'payments-queue':
+        const paymentStatus = successfull
+          ? PaymentStatus.SUCCEEDED
+          : PaymentStatus.FAILED;
+        const paymentDto: UpdatePaymentStatusDto = {
+          paymentId: msg.paymentId,
+          paymentStatus,
+        };
+        this.connectorService.sendUpdateToPayment(paymentDto);
+        break;
+
+      case 'shipments-queue':
+        const shipmentStatus = successfull
+          ? ShipmentStatus.DELIVERED
+          : ShipmentStatus.FAILED;
+        const shipmentDto: UpdateShipmentStatusDto = {
+          shipmentId: msg.shipmentId,
+          shipmentStatus,
+        };
+        this.connectorService.sendUpdateToShipment(shipmentDto);
+        break;
+    }
+  }
+
+  /**
+   * Resets all rate limits every minute
+   */
+  @Cron('0 * * * * *')
+  resetMessageCounts() {
+    this.logger.debug(`Resetting message counts.`);
+    this.queues.forEach((queue) => {
+      // check if processing was stopped for any queue
+      if (!this.processingAllowed[queue]) {
+        // close channels so Queue resends all unacknowledged messages
+        this.reconnectToRabbitMQ();
```

Comment:
this accesses this.channel (and sets it), however is not awaited
if there are multiple queues, could this result in async race conditions?

Reply:
Thanks for pointing this part out, It will reonnect multiple times, if multiple limits are reached. It should only reconnect once, since it reconnects for all queues anyway.